### DISCUSSION TITLE
Specify type of useLocale

### DISCUSSION THREAD

**Discussion by TommySorensen** [amannn/next-intl#477] [Q&A] [answered]
Anyone knows if there is a way to actually have a typesafe locale. Currently its just `string` but i want it to be my list of locales.

```js
// app/[locale]/layout.ts
export function LocaleLayout({ params }: Params) {

}

interface Params {
  params: { locale: Locale };
}

// Locale types
export enum Locales {
  en = 'en',
  enGb = 'en-GB'
}

export type Locale = (typeof localeConfig)['locales'][number];
```

<img width="429" alt="image" src="https://github.com/amannn/next-intl/assets/4134473/d8459ed4-b18c-4fcc-a220-b76848fa5312">

**Comment by amannn** [comment, marked as answer]
Currently only if you wrap the function:

```tsx
import {useLocale as useUntypedLocale} from 'next-intl';

export default function useLocale() {
  return useUntypedLocale() as 'en' | 'en-GB';
}
```

I've been thinking about a way to provide this within the library, as with the App Router, we can guarantee that only known locales are returned. For the Pages Router this was not possible with the way the library is integrated.

**Reply by TommySorensen** [reply]
Great. Thx for helping. This will work for now 🎉

What about making it possible to extend the interface like the `IntlMessages` example in the typescript section in the docs? That will only fix the types, but not the checking for known locales.

**Reply by amannn** [reply]
That would be an option, yes. In https://github.com/amannn/next-intl/pull/426 I'm working on localized pathnames by using a factory function, that'd be another option to return a typed `useLocale` hook. I need to give it a bit more thought, also considering other APIs so that we have a cohesive story.

**Reply by stefanprobst** [reply]
maybe something similar to what `vue-i18n` does could work? see: https://github.com/intlify/vue-i18n/pull/1890

**Reply by maxsv01** [reply]
any update?

**Reply by amannn** [reply]
Strictly-typed locales shipped in v4, please see the blog post!
